### PR TITLE
[1.x] Introduce strictTemplatePolicy

### LINKED ISSUE #5323: Introduce strictTemplatePolicy
### Description
If a developer allows untrusted & unsanitized injection of arbitrary HTML into the document (i.e. via `innerHTML`), expression evaluation in Polymer templates injected into the document may be used to achieve arbitrary code execution.

An opt-in `strictTemplatePolicy` mode should disable evaluation of any Polymer template that is not associated with a trusted element registered via script (which must always be trusted).

In general, enabling `strictTemplatePolicy` should cause the following changes in behavior:
* All "self-stamping" templates including `<dom-bind>`, `<dom-repeat>`, and `<dom-if>` should be prohibited if not originating from the template of a trusted element registered via script, to prevent evaluation of injected templates
* Re-registration of `<dom-module>` should be prohibited to disallow use of templates from injected `<dom-module>`s
* Elements that do not have templates by design must explicitly indicate a null template to prevent template retrieval from inject `<dom-module>`s, or else cause an error

### PR DESCRIPTION
**Pull request by kevinpschaaf**
Enabling `strictTemplatePolicy` by setting `Polymer = {setStrictTemplatePolicy: true};` prior to loading Polymer causes the following changes in behavior:
* All "self-stamping" templates including `<dom-bind>`, `<dom-repeat>`, and `<dom-if>` will throw errors and not stamp if not originating from inside the template of a trusted element registered via script
* Re-registration of `<dom-module>` will throw and cause template lookup for that element name to return `null`
* Elements that do not have templates by design must explicitly indicate a null template to prevent template retrieval from inject `<dom-module>`s, otherwise they will throw an error

Fixes #5323

### CODE REVIEW

File: src/lib/template/dom-bind.html [hunk 1]
```diff
@@ -78,6 +78,9 @@
     _template: null,
 
     created: function() {
+      if (Polymer.Settings.strictTemplatePolicy) {
```
Comment (sorvell) [inline src/lib/template/dom-bind.html]: Test this throws as expected using the custom elements polyfill (and doesn't halt other elements from booting up)

File: test/unit/strict-template-policy.html [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        assertThrows(function() {
+          document.getElementById('target').innerHTML =
+            '<template is="dom-repeat" items="[0]">' +
+            '  <div id="injected"></div>'+
+            '</template>`';
+        }, /template owner not trusted/);
+        assert.notOk(document.getElementById('injected'));
+      });
+
+      test('dom-module after registration', function() {
```
Comment (sorvell) [inline test/unit/strict-template-policy.html]: Add a test that registers the same module name 3x.

Comment (kevinpschaaf) [reply test/unit/strict-template-policy.html]: Added: https://github.com/Polymer/polymer/pull/5303/commits/1d529ab9512745457aa3873c6795624083658cc6#diff-e2ce14a791fe5ac94beadd7de8fcdc95R142

File: src/lib/template/dom-bind.html (hunk 1 above)
Comment (kevinpschaaf) [reply src/lib/template/dom-bind.html]: Confirmed that on CE polyfill (both v0 and v1) element upgrading halts when a constructor throws.

In V1 it actually try/catch'es, but [re-throws the error](https://github.com/webcomponents/custom-elements/blob/master/src/CustomElementInternals.js#L285).  Not sure why.  Filed issue here: https://github.com/webcomponents/custom-elements/issues/176

File: src/standard/annotations.html [hunk 3]
```diff
@@ -236,7 +236,7 @@
 
     // push configuration references at configure time
     _configureAnnotationReferences: function() {
-      var notes = this._notes;
+      var notes = this._notes || [];
```
Comment (sorvell) [inline src/standard/annotations.html]: Why is this changed needed?

File: src/mini/template.html [hunk 4]
```diff
@@ -29,7 +29,11 @@
       var module;
       if (this._template === undefined) {
         module = Polymer.DomModule.import(this.is);
-        this._template = module && module.querySelector('template');
+        var template = module && module.querySelector('template');
```
Comment (sorvell) [inline src/mini/template.html]: Isn't it incorrect to do this if `this._template` is explicitly `null`?

Comment (kevinpschaaf) [reply src/mini/template.html]: Correct, this is all skipped when `this._template === undefined`, including when template is explicitly set to null.

File: src/standard/annotations.html (hunk 3 above)
Comment (kevinpschaaf) [reply src/standard/annotations.html]: Good question.  When `_prepTemplate` throws under a `strictTemplatePolicy` violation, it leaves `_notes` uninitialized, which was previously not a case that could ever occur since `_prepAnnotations` comes after that and unconditionally sets `_notes`.  So this just provides a good fallback behavior under the violation case.
